Sources/SwiftSyntax/Identifier.swift

```diff
@@ -0,0 +1,60 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// An abstraction for sanitized values on a token.
+public struct Identifier: Equatable, Hashable, Sendable {
+  /// The sanitized `text` of a token.
+  public var name: String {
+    String(syntaxText: raw.name)
+  }
+
+  @_spi(RawSyntax)
+  public let raw: RawIdentifier
+
+  public init?(_ token: TokenSyntax) {
+    guard case .identifier = token.tokenKind else {
+      return nil
+    }
+
+    self.raw = RawIdentifier(token.tokenView)
+  }
+}
+
+@_spi(RawSyntax)
+public struct RawIdentifier: Equatable, Hashable, Sendable {
+  public let name: SyntaxText
+
+  private let arena: RetainedSyntaxArena
```

Comment:
The `Raw*` syntax nodes in swift-syntax shouldn’t retain their arenas, `RawSyntaxData` for example, also just has a `SyntaxArenaRef` instead of a `RetainedSyntaxArena`. The idea is that the user of the raw syntax nodes guarantees that the arena stays alive and thus doesn’t need to pay reference counting traffic.

In this concrete example, I think `RawIdentifier` should just store a `SyntaxText` and not have any arena reference (since it’s not needed) and `Identifier` is the safe type on top that makes sure the arena doesn’t get deallocated by retaining it through a `RetainedSyntaxArena` reference.

Reply:
The `testRawIdentifier` test references the `RawIdentifier`, and by moving the arena reference up to the Identifier its no longer retained and the test fails. Does testRawIdentifier need refactoring as well? I'm not sure what RawIdentifier can be used for without retaining the string that backs it?

Reply:
OK, as long as thats expected with SPI types then this works. I've updated the PR.

Reply:
It's expected for the raw syntax nodes, not really to do with SPI or not (it's just that they are SPI as they aren't the intended user-facing API).